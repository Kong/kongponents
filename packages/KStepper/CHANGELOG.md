# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 7.0.6 (2022-09-08)

**Note:** Version bump only for package @kongponents/kstepper





## 7.0.5 (2022-09-06)

**Note:** Version bump only for package @kongponents/kstepper





## 7.0.4 (2022-09-02)


### Bug Fixes

* **ktable:** client sorting with number values ([#775](https://github.com/Kong/kongponents/issues/775)) ([22efd29](https://github.com/Kong/kongponents/commit/22efd290a0f3954332ea2e8131520da519ae17b6))





## 7.0.3 (2022-09-01)


### Bug Fixes

* **kdropdownmenu:** add icon prop ([#771](https://github.com/Kong/kongponents/issues/771)) ([6526566](https://github.com/Kong/kongponents/commit/65265662247e59c1580640d0988986c2f9d7b185))





## 7.0.2 (2022-08-30)

**Note:** Version bump only for package @kongponents/kstepper





## 7.0.1 (2022-08-30)

**Note:** Version bump only for package @kongponents/kstepper





# [7.0.0](https://github.com/Kong/kongponents/compare/@kongponents/kstepper@7.0.0-alpha.2...@kongponents/kstepper@7.0.0) (2022-08-30)

**Note:** Version bump only for package @kongponents/kstepper





# 7.0.0-alpha.2 (2022-08-30)

**Note:** Version bump only for package @kongponents/kstepper





# [7.0.0-alpha.1](https://github.com/Kong/kongponents/compare/v6.7.6...v7.0.0-alpha.1) (2022-08-29)


### Bug Fixes

* **kstepper:** include all vue files in package ([#758](https://github.com/Kong/kongponents/issues/758)) ([1232cd1](https://github.com/Kong/kongponents/commit/1232cd14bc76556cfe4d110831736d7264eb1376))
* **kstepper:** include stepper-icons directory too ([#759](https://github.com/Kong/kongponents/issues/759)) ([e91c37b](https://github.com/Kong/kongponents/commit/e91c37be9fb5a8e0f86c7f590c35a70c75b76cae))
* **kstepper:** remove unused prop which is throwing errors ([#766](https://github.com/Kong/kongponents/issues/766)) ([f2825be](https://github.com/Kong/kongponents/commit/f2825bef268c5971b7c3e22b6eecae7dbd26cd44))


### Features

* **kstepper:** new component [khcp-4301] ([#745](https://github.com/Kong/kongponents/issues/745)) ([44c053d](https://github.com/Kong/kongponents/commit/44c053d9fad1a4b619ea4ef7b6cfa5cf4bbd9295))
* update vue for 2.7 [KHCP-4369] ([d9b5f27](https://github.com/Kong/kongponents/commit/d9b5f273693f52c40c6e43aa200d30a0cc4bd6a3))


### BREAKING CHANGES

* Update Kongponents Vue peerDependency to 2.7.x.





## 6.40.3 (2022-08-23)


### Reverts

* Revert "fix(kmodal): prevent click propagation [KHCP-4525] (#767)" ([f00261d](https://github.com/Kong/kongponents/commit/f00261d73308e3f836ea45d37827ee0ee0799795)), closes [#767](https://github.com/Kong/kongponents/issues/767)





## 6.40.2 (2022-08-22)


### Bug Fixes

* **kmodal:** prevent click propagation [KHCP-4525] ([#767](https://github.com/Kong/kongponents/issues/767)) ([e69502a](https://github.com/Kong/kongponents/commit/e69502afbcb12f1170ec848460382284f3b02035))





## 6.40.1 (2022-08-19)


### Bug Fixes

* **kstepper:** remove unused prop which is throwing errors ([#766](https://github.com/Kong/kongponents/issues/766)) ([f2825be](https://github.com/Kong/kongponents/commit/f2825bef268c5971b7c3e22b6eecae7dbd26cd44))





# 6.40.0 (2022-08-17)


### Features

* **kalert:** updates for KHCP-4405 ([#763](https://github.com/Kong/kongponents/issues/763)) ([6dadda8](https://github.com/Kong/kongponents/commit/6dadda87389bf45f09f7d171602d9cadd32c7a21))





## 6.39.1 (2022-08-17)

**Note:** Version bump only for package @kongponents/kstepper





# 6.39.0 (2022-08-17)


### Features

* **kselect:** add clearable prop ([#749](https://github.com/Kong/kongponents/issues/749)) ([f56e97f](https://github.com/Kong/kongponents/commit/f56e97f4f485451b5bd8e59ed114799b64fac2c8))





# 6.38.0 (2022-08-16)


### Features

* **icons:** add book, plug and contact support icons ([#760](https://github.com/Kong/kongponents/issues/760)) ([38af9da](https://github.com/Kong/kongponents/commit/38af9da57e80c776d88df79c25edc460a34cf1c2))





## 6.37.2 (2022-08-12)


### Bug Fixes

* **kstepper:** include stepper-icons directory too ([#759](https://github.com/Kong/kongponents/issues/759)) ([e91c37b](https://github.com/Kong/kongponents/commit/e91c37be9fb5a8e0f86c7f590c35a70c75b76cae))





## 6.37.1 (2022-08-12)


### Bug Fixes

* **kstepper:** include all vue files in package ([#758](https://github.com/Kong/kongponents/issues/758)) ([1232cd1](https://github.com/Kong/kongponents/commit/1232cd14bc76556cfe4d110831736d7264eb1376))





# 6.37.0 (2022-08-10)


### Features

* **kstepper:** new component [khcp-4301] ([#745](https://github.com/Kong/kongponents/issues/745)) ([44c053d](https://github.com/Kong/kongponents/commit/44c053d9fad1a4b619ea4ef7b6cfa5cf4bbd9295))
