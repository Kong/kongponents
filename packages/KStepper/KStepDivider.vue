<template>
  <svg
    width="214"
    height="2"
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      :fill="isCompleted ? completedColor : defaultColor"
      d="M0 0h214v2H0z"
    />
  </svg>
</template>

<script>
export default {
  name: 'KStepDivider',
  props: {
    defaultColor: {
      type: String,
      // TODO: var
      default: '#E7E7EC'
    },
    completedColor: {
      type: String,
      // TODO: var
      default: '#169FCC'
    },
    isCompleted: {
      type: Boolean,
      default: false
    }
  }
}
</script>
