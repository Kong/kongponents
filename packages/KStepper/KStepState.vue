<template>
  <div class="k-step-state mx-3">
    <KCompletedState v-if="state === 'completed'" />
    <KErrorState v-else-if="state === 'error'" />
    <KPendingState v-else-if="state === 'pending'" />
    <KDefaultState v-else />
  </div>
</template>

<script>
import KCompletedState from './stepper-icons/KCompletedState.vue'
import KDefaultState from './stepper-icons/KDefaultState.vue'
import KErrorState from './stepper-icons/KErrorState.vue'
import KPendingState from './stepper-icons/KPendingState.vue'

export default {
  name: 'KStepState',
  components: { KCompletedState, KDefaultState, KErrorState, KPendingState },
  props: {
    state: {
      type: String,
      default: 'default',
      validator: (value) => ['default', 'pending', 'completed', 'error'].includes(value)
    }
  }
}
</script>

<style lang="scss" scoped>
.k-step-state {
  position: relative;
  top: -8px;
}
</style>
