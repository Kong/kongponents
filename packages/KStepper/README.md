# @kongponents/kstepper

[![](https://img.shields.io/npm/v/@kongponents/kstepper.svg?style=flat-square)](https://www.npmjs.com/package/@kongponents/kstepper)

```html
<KStepper :description="'hello world'">
  Hello from a slot
</KStepper>
```
