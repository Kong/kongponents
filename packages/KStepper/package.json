{
  "name": "@kongponents/kstepper",
  "version": "6.36.6",
  "description": "An ordered Stepper component",
  "main": "dist/KStepper.umd.min.js",
  "componentName": "KStepper",
  "source": "KStepper.vue",
  "files": [
    "dist",
    "KStepper.vue"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@kongponents/klabel": "^6.36.6",
    "@kongponents/styles": "^6.36.6"
  }
}
