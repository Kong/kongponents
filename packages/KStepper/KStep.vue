<template>
  <div
    :class="{
      'is-first-step': isFirst,
      'is-last-step': isLast,
      'is-pending': state === 'pending'
    }"
    class="k-step"
  >
    <div class="k-step-state-container d-flex mb-5">
      <KStepState :state="state" />

      <div v-if="!isLast">
        <KStepDivider :is-completed="state === 'completed'" />
      </div>
    </div>

    <div
      :class="{ 'bolder': state === 'pending' || state === 'error' }"
      class="k-step-label"
    >
      <KLabel>
        {{ label }}
      </KLabel>
    </div>
  </div>
</template>

<script>
import KLabel from '@kongponents/klabel/KLabel.vue'
import KStepDivider from './KStepDivider.vue'
import KStepState from './KStepState.vue'

export default {
  name: 'KStep',
  components: { KLabel, KStepDivider, KStepState },
  props: {
    label: {
      type: String,
      required: true
    },
    state: {
      type: String,
      default: 'default',
      validator: (value) => ['default', 'pending', 'completed', 'error'].includes(value)
    },
    isFirst: {
      type: Boolean,
      default: false
    },
    isLast: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style lang="scss" scoped>
@import '~@kongponents/styles/variables';

.k-step {
  --KStepContainerWidth: 170px;
  --KStepStateWidth: 24px;

  width: fit-content;

  .k-step-label {
    --KInputLabelWeight: 400;

    width: var(--KStepContainerWidth);
    text-align: center;
    margin-left: calc(((#{var(--KStepContainerWidth)} * .5) - #{var(--KStepStateWidth)}) * -1);

    &.bolder {
      --KInputLabelWeight: 600;
    }
  }

  &.is-first-step {
    .k-step-state-container {
      margin-left: calc((#{var(--KStepContainerWidth)} * .5) - #{var(--KStepStateWidth)});
    }

    .k-step-label {
      margin-left: 0;
    }
  }
}
</style>
