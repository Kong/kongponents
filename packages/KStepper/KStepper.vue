<template>
  <ol class="k-stepper">
    <KStep
      v-for="(step) in steps"
      :key="`k-step-${step.label.replace(' ', '-')}`"
      :label="step.label"
      :state="step.state"
      :max-label-width="maxLabelWidth"
    />
  </ol>
</template>

<script>
import KStep from './KStep.vue'

export default {
  name: 'KStepper',
  components: { KStep },
  props: {
    /**
     * Array of steps to display
     */
    steps: {
      type: Array,
      required: true,
      validator: (items) => !items.length || items.some(i => Object.hasOwn(i, 'label'))
    },
    maxLabelWidth: {
      type: String,
      default: ''
    },
    /**
     * Test mode - for testing only, strips out generated ids
     */
    testMode: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style lang="scss" scoped>
.k-stepper {
  list-style: none;
  padding-inline-start: 0px;
  width: 100%;
  max-width: 100%;
  overflow-x: auto;
  display: flex;
  position: relative;
}
</style>
