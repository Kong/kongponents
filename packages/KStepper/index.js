import KStepper from './KStepper'
import KStep from './KStep'

const Plugin = {
  install (Vue) {
    Vue.component('KStepper', KStepper)
    Vue.component('KStep', KStep)
  }
}

export default Plugin
export { KStepper, KStep }

if (typeof window !== 'undefined' && window.Vue) {
  window.Vue.use(Plugin)
}
