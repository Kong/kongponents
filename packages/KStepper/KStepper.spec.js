import { mount } from '@vue/test-utils'
import KStepper from '@/KStepper/KStepper'

const longSteps = [
  { label: 'Step  a long long long long time ago', state: 'completed' },
  { label: 'in a galaxy far far away', state: 'completed' },
  { label: 'Kongponents were battling in space and', state: 'pending' },
  { label: 'fighting robots and space monsters with lots of explosions' }
]

const stepTypes = [
  { label: 'step: completed', state: 'completed' },
  { label: 'step: pending', state: 'pending' },
  { label: 'step: error', state: 'error' },
  { label: 'step: default' }
]

describe('KStepper', () => {
  it('renders steps', () => {
    const wrapper = mount(KStepper, {
      propsData: {
        steps: longSteps
      }
    })

    expect(wrapper.find('.k-stepper').exists()).toBeTruthy()
    expect(wrapper.findAll('.k-step')).toHaveLength(longSteps.length)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders with maxLabelWidth', () => {
    const width = 100

    const wrapper = mount(KStepper, {
      propsData: {
        maxLabelWidth: width + '',
        steps: longSteps
      }
    })

    const stepLabel = wrapper.find('.k-step-label')

    expect(wrapper.find('.k-stepper').exists()).toBeTruthy()
    expect(stepLabel.element.style['max-width']).toEqual(width + 'px')
  })

  it('renders each step type', () => {
    const wrapper = mount(KStepper, {
      propsData: {
        steps: stepTypes
      }
    })
    const types = ['completed', 'default', 'error', 'pending']

    expect(wrapper.find('.k-stepper').exists()).toBeTruthy()

    for (let i = 0; i < types.length; i++) {
      expect(wrapper.find(`[data-testid="k-step-${types[i]}"]`).exists()).toBeTruthy()
      expect(wrapper.find(`[data-testid="k-step-${types[i]}"]`).html()).toEqual(expect.stringContaining(types[i]))
    }

    expect(wrapper.html()).toMatchSnapshot()
  })
})
