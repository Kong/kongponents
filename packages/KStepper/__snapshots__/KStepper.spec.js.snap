// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KStepper renders each step type 1`] = `
<ol class="k-stepper">
  <li class="k-step">
    <div class="k-step-container completed">
      <div class="k-step-state px-3"><svg width="24" height="24" viewBox="0 0 24 24" fill="none" data-testid="k-step-completed">
          <title>Completed</title>
          <circle fill="var(--teal-300)" cx="12" cy="12" r="12"></circle>
          <path fill="var(--teal-300)" d="M16 9L10 15L7 12.0001"></path>
          <path stroke="var(--white)" d="M16 9L10 15L7 12.0001" stroke-width="2.25" stroke-linecap="round" stroke-linejoin="round"></path>
        </svg></div>
      <div class="k-step-label" style="max-width: px;"><label class="k-input-label">
          step: completed
        </label></div>
    </div>
  </li>
  <li class="k-step">
    <div class="k-step-container">
      <div class="k-step-state px-3">
        <div class="k-step-spinner"><svg width="24" height="24" viewBox="0 0 50 50" data-testid="k-step-pending">
            Pending
            <circle cx="25" cy="25" r="20" fill="none" stroke-width="5"></circle></svg></div>
      </div>
      <div class="k-step-label bolder" style="max-width: px;"><label class="k-input-label">
          step: pending
        </label></div>
    </div>
  </li>
  <li class="k-step">
    <div class="k-step-container">
      <div class="k-step-state px-3"><svg width="24" height="24" viewBox="0 0 24 24" fill="none" data-testid="k-step-error">
          <title>Error</title>
          <circle fill="var(--red-500)" cx="12" cy="12" r="12"></circle>
          <path fill="var(--white)" fill-rule="evenodd" clip-rule="evenodd" d="M15.7397 9.51743C16.0868 9.1703 16.0868 8.60748 15.7397 8.26035C15.3925 7.91322 14.8297 7.91322 14.4826 8.26035L12 10.7429L9.51746 8.26035C9.17033 7.91322 8.60751 7.91322 8.26038 8.26035C7.91325 8.60748 7.91325 9.1703 8.26038 9.51743L10.7429 12L8.26035 14.4826C7.91322 14.8297 7.91322 15.3925 8.26035 15.7397C8.60748 16.0868 9.1703 16.0868 9.51743 15.7397L12 13.2571L14.4826 15.7397C14.8297 16.0868 15.3925 16.0868 15.7397 15.7397C16.0868 15.3925 16.0868 14.8297 15.7397 14.4826L13.2571 12L15.7397 9.51743Z"></path>
        </svg></div>
      <div class="k-step-label bolder error" style="max-width: px;"><label class="k-input-label">
          step: error
        </label></div>
    </div>
  </li>
  <li class="k-step">
    <div class="k-step-container">
      <div class="k-step-state px-3"><svg width="24" height="24" viewBox="0 0 24 24" fill="none" data-testid="k-step-default">
          <title>Default</title>
          <circle fill="var(--grey-300)" cx="12" cy="12" r="12"></circle>
        </svg></div>
      <div class="k-step-label" style="max-width: px;"><label class="k-input-label">
          step: default
        </label></div>
    </div>
  </li>
</ol>
`;

exports[`KStepper renders steps 1`] = `
<ol class="k-stepper">
  <li class="k-step">
    <div class="k-step-container completed">
      <div class="k-step-state px-3"><svg width="24" height="24" viewBox="0 0 24 24" fill="none" data-testid="k-step-completed">
          <title>Completed</title>
          <circle fill="var(--teal-300)" cx="12" cy="12" r="12"></circle>
          <path fill="var(--teal-300)" d="M16 9L10 15L7 12.0001"></path>
          <path stroke="var(--white)" d="M16 9L10 15L7 12.0001" stroke-width="2.25" stroke-linecap="round" stroke-linejoin="round"></path>
        </svg></div>
      <div class="k-step-label" style="max-width: px;"><label class="k-input-label">
          Step a long long long long time ago
        </label></div>
    </div>
  </li>
  <li class="k-step">
    <div class="k-step-container completed">
      <div class="k-step-state px-3"><svg width="24" height="24" viewBox="0 0 24 24" fill="none" data-testid="k-step-completed">
          <title>Completed</title>
          <circle fill="var(--teal-300)" cx="12" cy="12" r="12"></circle>
          <path fill="var(--teal-300)" d="M16 9L10 15L7 12.0001"></path>
          <path stroke="var(--white)" d="M16 9L10 15L7 12.0001" stroke-width="2.25" stroke-linecap="round" stroke-linejoin="round"></path>
        </svg></div>
      <div class="k-step-label" style="max-width: px;"><label class="k-input-label">
          in a galaxy far far away
        </label></div>
    </div>
  </li>
  <li class="k-step">
    <div class="k-step-container">
      <div class="k-step-state px-3">
        <div class="k-step-spinner"><svg width="24" height="24" viewBox="0 0 50 50" data-testid="k-step-pending">
            Pending
            <circle cx="25" cy="25" r="20" fill="none" stroke-width="5"></circle></svg></div>
      </div>
      <div class="k-step-label bolder" style="max-width: px;"><label class="k-input-label">
          Kongponents were battling in space and
        </label></div>
    </div>
  </li>
  <li class="k-step">
    <div class="k-step-container">
      <div class="k-step-state px-3"><svg width="24" height="24" viewBox="0 0 24 24" fill="none" data-testid="k-step-default">
          <title>Default</title>
          <circle fill="var(--grey-300)" cx="12" cy="12" r="12"></circle>
        </svg></div>
      <div class="k-step-label" style="max-width: px;"><label class="k-input-label">
          fighting robots and space monsters with lots of explosions
        </label></div>
    </div>
  </li>
</ol>
`;
